### ISSUE TITLE
FEAT: Show total number of pages in pagination

### ISSUE THREAD

**Issue by tgrelka** [beekeeper-studio/beekeeper-studio#744]
**Is your feature request related to a problem? Please describe.**
When looking through tables, sometimes I want to go to the last page, or a specific page when paginating. As of the latest update (1.13), the pagination of tables in the bottom bar has been significantly reduced, now only showing two buttons and a number field.

**Describe the solution you'd like**
I’d like to see the total number of pages, in a "X of 100" kind of way, e.g.
<  [ 1 ]  >   *of 100*
or maybe
<  [ 1 ]  >  *(100 pages)*

**Comment by rathboma** [comment]
Hey! Yes we used to have more pagination, but unfortunately relying on a table record count was killing performance for many production tables.

This is a stop-gap. We may be able to add total pages to unfiltered table views as we eventually load in the record count

**Comment by rathboma** [comment]
Note for self:
- Show total number of pages in the unfiltered view (totalRecords / pageSize)
- In the filtered view, just don't show the total number of pages

**Comment by tgrelka** [comment]
> We may be able to add total pages to unfiltered table views as we eventually load in the record count
@rathboma That’s what’s been confusing me — the record count is displayed, but not the total number of pages.
Actually, I didn’t have too many performance complaints before, working on a huge database with tens of millions of records.

I took a quick look at the code, and it seems like you’re using `count(*)` to get the total row count of a table. This can be a rather slow approach, since databases typically check each row individually.
I’ll try to free up some time in 2-3 weeks and see if I can build a more performant solution which also works well on filtered views. Even if it doesn’t, I suppose a good, and especially a quick estimate will also help.

BTW, I think some options might be good: Number of rows for each page (couldn’t find one) and offset in rows instead of pages. If I only manage to implement row count as an estimate a toggle for precision vs. performance would be nice — in combination with showing page count in filtered views, depending on the filter they tend to be A LOT smaller. (plus, sometimes you’re only really interested in the number of rows given a filter). I’ll split this into a new feature request very soon, promised.

**Comment by rathboma** [comment]
Hey! The different databases we actually have some other ways to approximate row count. Postgres SQL server and my SQL in particular.

Even with that some people's databases were falling over on these page loads so we decided it was better to have it perform well for everybody initially and then add backing features that don't impact that.

I agree there are many more paging options that need adding back in and configurable, frankly while these are useful they're not a big priority at the moment as we push out a few more bigger features.

It certainly seems like something that could be an easy addition for a first-time contributor. If you want to file another ticket with the specific paging options you think would be useful I can tag that as good for beginners

**Comment by tgrelka** [comment]
For me, paging is mostly the reason I use a graphical SQL client. Quick overview over data, how much, what (roughly), how many contain X, just some exploration of tables.
While building metrics to analyse usage is great, sometimes they’re just not enough, too inflexible or take too much time to implement when you can just have a quick gander at the tables in question.

I’ll likely be downgrading to 1.12 in the meantime, but see if I can find time myself to fix the regressions of the pagination. At the very least, I’ll open a ticket.

Thank you for your time! 😄

**Comment by MasterOdin** [comment]
For something like fetching statistics, you'll want to test against data warehouse providers (e.g. redshift) over postgres for performance measuring. Because of how they work, queries that miss the "cache" can be very costly due to how it then has to be compiled and executed, versus something like postgres proper.

**Comment by rathboma** [comment]
Yep. That plus poorly optimized or high traffic relational tables is where users saw problems.

**Comment by tgrelka** [comment]
Data warehousing certainly is effective, but also adds additional expense and requires maintenance to aggregate and run OLAP. 

I’m working in a small team, where right now it’s not efficient for us to integrate a data warehouse, especially given that due to high privacy constraints we avoid products, especially hosting, from providers like Google, Amazon and Microsoft like the plague.

In turn, the cost of running quick one-off analysis on a dump or just snooping around tables for twenty minutes occasionally is much better in getting us the results we want. (Oh, they’re using these fields for an unforeseen purpose. Ah, feature adoption hasn’t been that great so far. etc)
While in the long run, DW is the better option, it would require a lot of man-hours to get it into a usable state, which simply isn’t affordable right now.

So in summary, Beekeeper, or any other graphical SQL client, for that matter, is just an exploration tool for me, usually run on some local DB instance in a container. My example my have been a bad one, as I also use it to check the state of test databases for debugging purposes, though my point remains, proper pagination is one of the core features of a graphical DB client for me.

**Comment by rathboma** [comment]
@tgrelka you are exactly the target of the app! I built it to do exactly what you describe.

I was mentioning data warehouses as the reason for having slow count queries which caused us to change our UI a little :-). I wasn't in any way suggesting you should use one.

We encounter such varied database set-ups, we're trying super hard to make the app 'just work' for as many as possible, which unfortunately means that sometimes we have feature regressions as we learn that the app caused their database to misbehave.
